README.md

```diff
@@ -1 +1,158 @@
+
+Understanding asynchronous code and its patterns is the key to mastering Node.js. Besides the async code, and a few differences like module system, globals and data types, Node is almost the same as browser JavaScript. For this reason, let's dig deeper into the async code and how we can implement it.
+
+This lesson will cover the async code on the example of `fs`.
+
+## Objectives
+
+1.Illustrate async code wtih fs read method
+1. Illustrate async code wtih fs write method
+
+## Async Way of Thinking
+
+To warm up with the async code, consider an example in which we have a few `console.log` statements and a timeout:
+
+```js
+var a = 1
+console.log('The value of a is ', a)
+setTimeout(function(){
+  a = 2 
+  console.log('While the value of a is', a)
+}, 500)
+console.log('Now the value of a is', a)
+```
+
+For someone not versed in async way of thinking, it might appear that the resulting output will be 
+
+```
+The value of a is  1
+While the value of a is 2
+Now the value of a is 2
+```
+
+Because in most languages the execution goes in the order of the statements, a person might think that because `a =2` precedes `Now the value...`, the second and third output will have the values of 2. This is incorrect, because `setTimeout()` is scheduled asynchronously in the future. The right output will be:
+
+```js
+The value of a is  1
+Now the value of a is 1
+// delay of roughly 500 miliseconds
+While the value of a is 2
+```
+
+This misunderstanding of async code leads to many bugs especially for developers new to JavaScript and Node. That's why it's important to start thinking async way.
+
+
+## fs.readFile
+
+So to read a file  in the synchronous mode, we would use this method:
+
+```js
+var content = fs.readFileSync(filename, options)
+```
+
+The above code will block the execution of everything until the file is read. So if we add output:
+
+```js
+var fs = require('fs')
+console.log('Start reading file')
+var content = fs.readFileSync('readme.md', 'utf-8')
+console.log('Finish reading file')
+console.log(content)
+```
+
+The code will produce "Start reading file", then wait until the file reading is over and finally print "Finish reading file".
+
+The async method `fs.readFile` is a better because our program can do something else while it waits for the content of the file. There are a few things which are different in `fs.readFile` comparing to `fs.readFileSync`:
+
+1. We need to use callback
+2. The callback has an error object
+3. The callback has the content object (the data from the file)
+4. We don't get the content right away because `readFile` is **not an expression** like `fs.readFileSync`. (Expressions are functions which return something.) In other words, we cannot write `var content = fs.readFile(...)`.
```

Comment:
what does fs.readFile return? how can we use it  process data in a file if we can't store the contents?